lib/tool_shed/util/shed_util_common.py

```diff
@@ -604,36 +604,36 @@ def get_repository_for_dependency_relationship( app, tool_shed, name, owner, cha
-    if not repository:
-        tool_shed_url = common_util.get_tool_shed_url_from_tool_shed_registry( app, tool_shed )
-        repository_clone_url = os.path.join( tool_shed_url, 'repos', owner, name )
-        repo_info_tuple = (None, repository_clone_url, changeset_revision, None, owner, None, None)
-        repository, pcr = repository_was_previously_installed( app, tool_shed_url, name, repo_info_tuple )
-    if not repository:
-        # The received changeset_revision is no longer installable, so get the next changeset_revision
-        # in the repository's changelog in the tool shed that is associated with repository_metadata.
-        tool_shed_url = common_util.get_tool_shed_url_from_tool_shed_registry( app, tool_shed )
-        params = dict( name=name, owner=owner, changeset_revision=changeset_revision )
-        pathspec = [ 'repository', 'next_installable_changeset_revision' ]
-        text = util.url_get( tool_shed_url, password_mgr=app.tool_shed_registry.url_auth( tool_shed_url ), pathspec=pathspec, params=params )
-        if text:
-            repository = get_installed_repository( app=app,
-                                                   tool_shed=tool_shed,
-                                                   name=name,
-                                                   owner=owner,
-                                                   changeset_revision=text )
-    return repository
+    repositories = get_installed_repository( app=app,
+                                             tool_shed=tool_shed,
+                                             name=name,
+                                             owner=owner,
+                                             return_all=True)
+    if repositories:
+        if repositories[0].installed_changeset_revision == changeset_revision:
+            repository = repositories[0]
+        else:
+            changeset_repo_d = { repo.changeset_revision: repo for repo in repositories }
+            repository = changeset_repo_d.get( changeset_revision, None )
+        if not repository:
+            tool_shed_url = common_util.get_tool_shed_url_from_tool_shed_registry( app, tool_shed )
+            changeset_revisions = get_previous_changeset_revisions(app=app,
+                                                                   tool_shed_url=tool_shed_url,
+                                                                   name=name,
+                                                                   owner=owner,
+                                                                   changeset_revision=changeset_revision)
+            for previous_changeset in changeset_revisions:
+                if previous_changeset in changeset_repo_d:
+                    return changeset_repo_d[ previous_changeset ]
+            # The received changeset_revision is no longer installable, so get the next changeset_revision
```

Comment:
There is something falling through the cracks here:
Assuming the received changeset_revision is 2, and the installed (and current) changeset revision is 3, and the next installed changeset revision is 4 (because next_installable_changeset_revision returns the latest changeset revision for packages), we do not find the installed repository at changeset revision 3, even though I believe the logic is aiming for this.

This could be fixed by first asking for the next_changeset_revision, and then for previous_chanegset revision (relative to next_changeset_revision).
But if we do this, we might as well ignore the changeset_revision completely, and just return whatever package matches toolshed/name/owner at the highest ctx_revision from the database,
instead of sending requests to the toolshed.
Does anyone see any problem with either of those approaches? [I would obviously prefer the last]
Ping @martenson @davebx @gregvonkuster
If we go with the last approach we’ll probably have another significant speedup and reduction of code.

places where get_repository_for_dependency_relationship is used:

lib/tool_shed/metadata/metadata_generator.py —> handle_repository_elem()  would give nice speedup, effective behaviour anyway
lib/tool_shed/galaxy_install/installed_repository_manager.py —> get_dependencies_for_repository(): shouldn’t make any difference, or slightly more robust to changes in recent dependencies ?!
lib/tool_shed/galaxy_install/tool_dependencies/env_manager.py -> get_env_shell_file_paths(): will find latest package. more reliable, I guess ? 
lib/tool_shed/galaxy_install/tool_dependencies/recipe/tag_handler.py -> handle_complex_repository_dependency_for_package () -> would fix the problem mentioned above
lib/tool_shed/util/shed_util_common.py ->get_repository_ids_requiring_prior_import_or_install() : this is the effective behaviour anyway, reduction of code